proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1269 @@
+
+   * ``[]`` is always an empty list
+   * ``()`` is always the unit data constructor
+   * ``[a]`` is always a singleton list (not the list type)
+   * ``(,)`` is always the pair data constructor (not the type constructor)
+   * ``(a, b)`` is always a pair (not the type of a pair)
+
+   Likewise for tuples of higher arities.
+
+   Export the following synonym from the ``Data.List`` module::
+
+     type List = []
+
+   Export the following synonyms from the ``Data.Tuple`` module::
+
+     type Unit = ()
+     type TupleOf2 = (,)
+     type TupleOf3 = (,,)
+     type TupleOf4 = (,,,)
+     ... -- up to the maximum tuple arity
+
+   This change allows the use of built-in lists and tuples without any
+   disambugation syntax (the ``'`` promotion syntax at the type level or the
+   ``type`` herald at the term level).
+
+4. When ``ViewPatterns`` are enabled, interpret ``f (a -> b) = ...``
+   as a view pattern, otherwise as ``f ((->) a b) = ...``.
+
+5. ``case ... of x -> y -> z`` is an error. We require parentheses to
+   disambiguate:
+
+   * ``case ... of (x -> y) -> z``
+   * ``case ... of x -> (y -> z)``
+
+Name resolution
+~~~~~~~~~~~~~~~
+
+6. During name resolution,
+
+   * Identifiers bound in term syntax populate the term namespace;
+     identifiers bound in type syntax populate the type namespace.
+
+     This is already the case, but now we generalize this rule to cover
+     types-in-terms, which are considered term syntax.
+
+   * When looking up an identifier ``v`` or ``V`` in type syntax, look it up
+     in the type namespace first; if it is not found there, look it up in the
+     term namespace.
+
+     This is already the case for uppercase identifiers if ``DataKinds`` is
+     enabled, but now we extend this rule to lowercase identifiers if
+     ``RequiredTypeArguments`` is enabled.
+
+   * When looking up an identifier ``v`` or ``V`` in term syntax, look it up
+     in the term namespace first; if it is not found there, look it up in the
+     type namespace.
+
+     This is a new rule, but notice how it mirrors the one for type syntax.
+
+Implicit quantification
+~~~~~~~~~~~~~~~~~~~~~~~
+
+7. Implicit quantification is an existing feature that allows the programmer to
+   omit a ``forall``::
+
+     g ::           a -> a    -- implicit
+     g :: forall a. a -> a    -- explicit
+
+   This sort of quantification only happens if the variable is not already in
+   scope::
+
+     {-# LANGUAGE ScopedTypeVariables #-}
+
+     f :: forall a. a -> a
+     f = ...
+       where
+         g :: a -> a         -- No implicit quantification!
+
+   In other words, we quantify only over *free* variables.
+
+   With the proposed changes to name resolution, variables that were previously
+   free are not free anymore::
+
+       a = 42
+       f :: a -> a           -- No implicit quantification!
+
+   This is a breaking change, and that is why the fallback to the term
```

Comment:
Ooh. I don't like this, but it is unavoidable.

Maybe a warning for `-Wcompat` even when `RequiredTypeArguments` is off?